File: android/core/ui/src/main/kotlin/com/devora/mencare/core/ui/notifications/NotificationsScreen.kt
package com.devora.mencare.core.ui.notifications

import androidx.compose.foundation.background
import androidx.compose.foundation.border
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.PaddingValues
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.lazy.LazyColumn
import androidx.compose.foundation.lazy.items
import androidx.compose.foundation.shape.CircleShape
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.automirrored.filled.ArrowBack
import androidx.compose.material3.Icon
import androidx.compose.material3.IconButton
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.getValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.res.stringResource
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import androidx.hilt.navigation.compose.hiltViewModel
import androidx.lifecycle.compose.collectAsStateWithLifecycle
import com.devora.mencare.core.common.formatDateTime
import com.devora.mencare.core.designsystem.R as DsR
import com.devora.mencare.core.designsystem.component.BrandSectionLabel
import com.devora.mencare.core.designsystem.component.DarkHeader
import com.devora.mencare.core.designsystem.component.DarkHeaderTitle
import com.devora.mencare.core.designsystem.component.readableWidth
import com.devora.mencare.core.designsystem.theme.Bone
import com.devora.mencare.core.designsystem.theme.Ink
import com.devora.mencare.core.designsystem.theme.Meta
import com.devora.mencare.core.designsystem.theme.OliveWood
import com.devora.mencare.core.designsystem.theme.Radii
import com.devora.mencare.core.designsystem.theme.StoneBorder
import com.devora.mencare.core.designsystem.theme.StoneSoft
import com.devora.mencare.core.designsystem.theme.TextMuted
import com.devora.mencare.core.ui.R
import java.time.LocalDate

/** Notification list: the same page for client, staff and owner. */
@Composable
fun NotificationsScreen(
    onBack: () -> Unit,
    viewModel: NotificationsViewModel = hiltViewModel(),
) {
    val state by viewModel.state.collectAsStateWithLifecycle()

    Column(modifier = Modifier.fillMaxSize().background(Bone)) {
        DarkHeader {
            Row(verticalAlignment = Alignment.CenterVertically) {
                IconButton(onClick = onBack) {
                    Icon(
                        Icons.AutoMirrored.Filled.ArrowBack,
                        contentDescription = stringResource(DsR.string.ds_back),
                        tint = Bone,
                    )
                }
                DarkHeaderTitle(title = stringResource(R.string.notifications_title))
            }
        }
        if (state.notifications.isEmpty()) {
            Text(
                stringResource(R.string.notifications_empty),
                style = MaterialTheme.typography.bodyMedium,
                color = MaterialTheme.colorScheme.onSurfaceVariant,
                modifier = Modifier.readableWidth().padding(20.dp),
            )
        } else {
            LazyColumn(
                modifier = Modifier.readableWidth(),
                contentPadding = PaddingValues(20.dp),
                verticalArrangement = Arrangement.spacedBy(10.dp),
            ) {
                // Le due letture della pagina — "da leggere" e "già viste" —
                // hanno due superfici, non due tonalità dello stesso pallino:
                // la card con il filo per le nuove, il grigio tenue per le
                // altre, con la data a separarle come nel mockup.
                val today = LocalDate.now()
                val groups = state.notifications.groupBy { it.at.toLocalDate() == today }
                listOf(true to R.string.notifications_today, false to R.string.notifications_earlier)
                    .forEach { (isToday, labelRes) ->
                        val rows = groups[isToday].orEmpty()
                        if (rows.isEmpty()) return@forEach
                        item(key = "header_$isToday") {
                            BrandSectionLabel(
                                stringResource(labelRes),
                                modifier = Modifier.padding(top = 4.dp, bottom = 2.dp),
                            )
                        }
                        items(rows, key = { it.id }) { notification ->
                            val isNew = notification.id in state.newIds
                            NotificationRow(notification.title, notification.body, formatDateTime(notification.at), isNew)
                        }
                    }
            }
        }
    }
}

/**
 * Una riga della campanella. Non letta: card chiara col filo e il punto
 * d'accento. Già letta: fondo tenue, nessun filo e il punto spento — la
 * differenza si vede dalla superficie, non solo da un pallino più chiaro.
 */
@Composable
private fun NotificationRow(title: String, body: String, at: String, isNew: Boolean) {
    Row(
        modifier = Modifier
            .fillMaxWidth()
            .clip(Radii.Md)
            .background(if (isNew) Bone else StoneSoft)
            .then(if (isNew) Modifier.border(1.5.dp, StoneBorder, Radii.Md) else Modifier)
            .padding(14.dp),
        verticalAlignment = Alignment.CenterVertically,
    ) {
        Box(
            Modifier
                .padding(end = 12.dp)
                .size(9.dp)
                .background(if (isNew) OliveWood else StoneBorder, CircleShape),
        )
        Column {
            Text(
                title,
                style = if (isNew) {
                    MaterialTheme.typography.titleSmall.copy(fontSize = 15.sp)
                } else {
                    MaterialTheme.typography.bodyLarge.copy(fontSize = 15.sp)
                },
                color = Ink,
            )
            Text(body, style = MaterialTheme.typography.bodyMedium, color = TextMuted)
            Text(at, style = Meta, color = TextMuted)
        }
    }
}

File: ios/MenCare/Core/UI/Notifications/NotificationsScreen.swift
import SwiftUI
import Observation

/// La campanella aveva la sua lettura dentro `body` e perfino dentro `init`:
/// con una lettura di rete non era piu' possibile, e serviva comunque uno stato
/// di caricamento. Adesso ha un view model come tutte le altre schermate.
@MainActor
@Observable
final class NotificationsViewModel {

    private let repository: NotificationRepository

    let state = LoadState()
    private(set) var notifications: [AppNotification] = []
    /// Quelle non lette quando la pagina si e' aperta: tengono il pallino finche'
    /// l'utente non esce, anche dopo che il server le ha segnate lette.
    private(set) var newIds: Set<String> = []

    init(repository: NotificationRepository) {
        self.repository = repository
    }

    func load() async {
        await state.run {
            let feed = try await repository.feed()
            notifications = feed.notifications
            if newIds.isEmpty {
                newIds = Set(feed.notifications.filter { !$0.read }.map(\.id))
            }
        }
        // Aprire la pagina vale come "viste": il contatore della campanella si
        // azzera sul server, non solo a schermo.
        _ = await repository.markAllRead()
    }
}

/// Lista delle notifiche: la stessa pagina per cliente, operatore e titolare.
struct NotificationsScreen: View {
    @State var viewModel: NotificationsViewModel
    let onBack: () -> Void

    var body: some View {
        VStack(spacing: 0) {
            DarkHeader {
                HStack(spacing: 4) {
                    Button(action: onBack) {
                        Image(systemName: "chevron.left")
                            .font(.system(size: 17, weight: .medium))
                            .foregroundStyle(Color.bone)
                            .padding(8)
                    }
                    .buttonStyle(.plain)
                    .accessibilityLabel(L("ds_back"))
                    DarkHeaderTitle(title: L("notifications_title"))
                }
            }
            Loadable(state: viewModel.state, retry: { Task { await viewModel.load() } }) {
                list
            }
        }
        .background(Color.bone)
        .task { await viewModel.load() }
    }

    @ViewBuilder
    private var list: some View {
        if viewModel.notifications.isEmpty {
            VStack(spacing: 0) {
                Text(L("notifications_empty"))
                    .font(Typo.bodyMedium)
                    .foregroundStyle(Color.textMuted)
                    .frame(maxWidth: .infinity, alignment: .leading)
                    .padding(20)
                    .readableWidth()
                Spacer()
            }
        } else {
            ScrollView {
                VStack(alignment: .leading, spacing: 10) {
                    // Le due letture della pagina — "da leggere" e "già viste"
                    // — hanno due superfici, non due tonalità dello stesso
                    // pallino, con la data a separarle come nel mockup.
                    let today = LocalDate.today()
                    let groups = Dictionary(grouping: viewModel.notifications) { $0.at.date == today }
                    ForEach([true, false], id: \.self) { isToday in
                        let rows = groups[isToday] ?? []
                        if !rows.isEmpty {
                            BrandSectionLabel(text: L(isToday ? "notifications_today" : "notifications_earlier"))
                                .padding(.top, 4)
                            ForEach(rows) { notification in
                                row(notification)
                            }
                        }
                    }
                }
                .padding(20)
                .readableWidth()
            }
        }
    }

    /// Una riga della campanella. Non letta: card chiara col filo e il punto
    /// d'accento. Già letta: fondo tenue, nessun filo e il punto spento — la
    /// differenza si vede dalla superficie, non solo da un pallino più chiaro.
    private func row(_ notification: AppNotification) -> some View {
        let isNew = viewModel.newIds.contains(notification.id)
        return HStack(alignment: .center, spacing: 12) {
            Circle()
                .fill(isNew ? Color.oliveWood : Color.stoneBorder)
                .frame(width: 9, height: 9)
            VStack(alignment: .leading, spacing: 2) {
                Text(notification.title)
                    .font(isNew ? Typo.titleSmall : Typo.bodyLarge)
                    .foregroundStyle(Color.ink)
                Text(notification.body)
                    .font(Typo.bodyMedium)
                    .foregroundStyle(Color.textMuted)
                Text(formatDateTime(notification.at))
                    .font(Typo.labelSmall)
                    .foregroundStyle(Color.textMuted)
            }
            .frame(maxWidth: .infinity, alignment: .leading)
        }
        .padding(14)
        .background(RoundedRectangle(cornerRadius: Radii.md).fill(isNew ? Color.bone : Color.stoneSoft))
        .overlay(
            RoundedRectangle(cornerRadius: Radii.md)
                .strokeBorder(Color.stoneBorder, lineWidth: isNew ? 1.5 : 0)
        )
    }
}
